src/node.cc

```diff
@@ -1303,86 +1303,187 @@ void AddPromiseHook(v8::Isolate* isolate, promise_hook_func fn, void* arg) {
+                        const async_context& asyncContext);
+  ~InternalCallbackScope();
+  void Close();
 
-  Environment::AsyncCallbackScope callback_scope(env);
-  bool disposed_domain = false;
+  inline bool Failed() const { return failed_; }
+  inline void MarkAsFailed() { failed_ = true; }
+  inline v8::Local<v8::Value> GetFailureReturnValue() const {
+    return failure_return_value_;
+  }
+  inline bool IsInnerMakeCallback() const {
+    return callback_scope_.in_makecallback();
+  }
 
-  if (recv->IsObject()) {
-    object = recv.As<Object>();
+  inline void UseTryCatch() {
+    CHECK_EQ(try_catch_, nullptr);
+    try_catch_ = new _TryCatch(env_->isolate());
+    try_catch_->real_try_catch.SetVerbose(true);
   }
 
+ private:
+  Environment* env_;
+  async_context async_context_;
+  v8::Local<v8::Object> object_;
+  Environment::AsyncCallbackScope callback_scope_;
+  bool failed_ = false;
+  bool pushed_ids_ = false;
+  bool closed_ = false;
+  _TryCatch* try_catch_ = nullptr;
+
+  // This is only for backwards compatibility, semantically it's always empty.
+  v8::Local<v8::Value> failure_return_value_;
+};
+
+CallbackScope::CallbackScope(Isolate* isolate,
+                             Local<Object> object,
+                             async_context asyncContext)
+  : private_(new InternalCallbackScope(Environment::GetCurrent(isolate),
+                                       object,
+                                       asyncContext)) {
+  private_->UseTryCatch();
+}
+
+CallbackScope::~CallbackScope() {
+  delete private_;
+}
+
+InternalCallbackScope::InternalCallbackScope(Environment* env,
+                                             Local<Object> object,
+                                             const async_context& asyncContext)
+  : env_(env),
+    async_context_(asyncContext),
+    callback_scope_(env) {
+  CHECK(!object.IsEmpty());
+  if (object->IsObject())
+    object_ = object.As<Object>();
+
+  // If you hit this assertion, you forgot to enter the v8::Context first.
+  CHECK_EQ(env->context(), env->isolate()->GetCurrentContext());
+
   if (env->using_domains()) {
-    CHECK(recv->IsObject());
-    disposed_domain = DomainEnter(env, object);
-    if (disposed_domain) return Undefined(env->isolate());
+    CHECK(object->IsObject());
+    failed_ = DomainEnter(env, object_);
+    if (failed_) {
+      failure_return_value_ = Undefined(env->isolate());
+      return;
+    }
   }
 
-  MaybeLocal<Value> ret;
+  if (asyncContext.async_id != 0) {
+    // No need to check a return value because the application will exit if
+    // an exception occurs.
+    AsyncWrap::EmitBefore(env, asyncContext.async_id);
+  }
 
-  {
-    AsyncHooks::ExecScope exec_scope(env, asyncContext.async_id,
-                                     asyncContext.trigger_async_id);
+  env->async_hooks()->push_ids(async_context_.async_id,
+                               async_context_.trigger_async_id);
+  pushed_ids_ = true;
```

Comment:
I don't see how `pushed_ids_` in `Close()` could be false. Mind clarifying?

Reply:
`pushed_ids_` is only true if the `InternalCallbackScope` constructor finishes, but that doesn’t happen if `DomainEnter()` reported that we’re trying to enter a disposed domain.